src/main/scala/org/graphframes/lib/AggregateMessages.scala

```diff
@@ -0,0 +1,161 @@
+     * executed on it.
+     *
+     * This returns a DataFrame with schema:
+     *  - column "id": vertex ID
+     *  - aggCol: aggregate result
+     * If you need to join this with the original [[GraphFrame.vertices]], you can run an inner
+     * join of the form:
+     * {{{
+     *   val g: GraphFrame = ...
+     *   val aggResult = g.aggregateMessages.sendToSrc(msg).agg(aggFunc)
+     *   aggResult.join(g.vertices, ID)
+     * }}}
+     */
+    def agg(aggCol: Column): DataFrame = {
+      require(msgToSrc.nonEmpty || msgToDst.nonEmpty, s"To run GraphFrame.aggregateMessages," +
+        s" messages must be sent to src, dst, or both.  Set using sendToSrc(), sendToDst().")
+      val triplets = g.triplets
+      import org.graphframes.GraphFrame.{DST, ID, SRC}
+      val sentMsgsToSrc = msgToSrc.map { msg =>
+        val msgsToSrc = triplets.select(msg.as("MSG"), triplets(SRC)(ID).as(ID))
+        // Inner join: only send messages to vertices with edges
+        msgsToSrc.join(g.vertices, ID)
+          .select(msgsToSrc("MSG"), col(ID))
+      }
+      val sentMsgsToDst = msgToDst.map { msg =>
+        val msgsToDst = triplets.select(msg.as("MSG"), triplets(DST)(ID).as(ID))
+        msgsToDst.join(g.vertices, ID)
+          .select(msgsToDst("MSG"), col(ID))
+      }
+      val unionedMsgs = (sentMsgsToSrc, sentMsgsToDst) match {
+        case (Some(toSrc), Some(toDst)) =>
+          toSrc.unionAll(toDst)
+        case (Some(toSrc), None) => toSrc
+        case (None, Some(toDst)) => toDst
+        case _ =>
+          // Should never happen. Specify this case to avoid compilation warnings.
+          throw new RuntimeException("AggregateMessages: No messages were specified to be sent.")
+      }
+      unionedMsgs.groupBy(ID).agg(aggCol)
+    }
+  }
+
+  /** Reference for source column, used for specifying messages */
+  def src: Column = col(GraphFrame.SRC)
+
+  /** Reference for destination column, used for specifying messages */
+  def dst: Column = col(GraphFrame.DST)
+
+  /** Reference for edge column, used for specifying messages */
+  def edge: Column = col(GraphFrame.EDGE)
+
+  /** Reference for message column, used for specifying aggregation function */
+  def msg: Column = col("MSG")
+
+  /**
+   * Cache a DataFrame, and returned the cached version. For iterative DataFrame-based algorithms.
+   *
+   * WARNING: This is NOT the same as [[DataFrame.cache()]].
+   *          The original DataFrame will NOT be cached.
+   *
+   * This is a workaround for SPARK-13346, which makes it difficult to use DataFrames in iterative
+   * algorithms.  This workaround converts the DataFrame to an RDD, caches the RDD, and creates
+   * a new DataFrame.  This is important for avoiding the creation of extremely complex DataFrame
+   * query plans when using DataFrames in iterative algorithms.
+   */
+  def getCachedDataFrame(df: DataFrame): DataFrame = {
```

Comment:
You should mention eplicitly this function should not be considered a stable api